server/rest/src/main/java/org/infinispan/rest/resources/CounterResource.java

```diff
@@ -30,84 +50,299 @@
+      return executeWeakCounterOp(request, WeakCounter::decrement);
+   }
+
+   private CompletionStage<RestResponse> addValueWeak(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
       String counterName = request.variables().get("counterName");
-      NettyRestResponse.Builder builder = new NettyRestResponse.Builder();
+      Long delta = checkForNumericParam("delta", request, responseBuilder);
+      if (delta == null) return completedFuture(responseBuilder.build());
 
-      CounterConfiguration configuration = counterManager.getConfiguration(counterName);
-      if (configuration == null) {
-         return CompletableFuture.completedFuture(builder.status(HttpResponseStatus.NOT_FOUND.code()).entity("Counter not found").build());
-      }
-      builder.header(HttpHeaderNames.CACHE_CONTROL.toString(), CacheControl.noCache());
+      CompletionStage<WeakCounter> weakCounter = getWeakCounter(counterName);
+      return weakCounter.thenCompose(counter -> {
+         if (counter == null) {
+            responseBuilder.status(HttpResponseStatus.BAD_REQUEST).entity(String.format("Weak counter '%s' not found", counterName));
+            return completedFuture(responseBuilder.build());
+         }
+         counter.add(delta);
+         return completedFuture(responseBuilder.build());
+      });
+   }
 
-      ContentSource contents = request.contents();
-      long delta;
-      if (contents == null) {
-         delta = 1;
-      } else {
-         String s = contents.asString();
-         delta = Long.parseLong(s);
-      }
-      if (configuration.type() == CounterType.WEAK)
-         return counterManager.getWeakCounter(counterName).add(delta).thenApply(v -> builder.build());
+   private CompletionStage<RestResponse> incrementStrong(RestRequest request) {
+      return executeStrongCounterOp(request, StrongCounter::incrementAndGet);
+   }
 
-      return counterManager.getStrongCounter(counterName).addAndGet(delta).thenApply(value -> builder.entity(Long.toString(value)).build());
+   private CompletionStage<RestResponse> decrementStrong(RestRequest request) {
+      return executeStrongCounterOp(request, StrongCounter::decrementAndGet);
    }
 
-   private CompletionStage<RestResponse> resetCounter(RestRequest request) throws RestResponseException {
+   private CompletionStage<RestResponse> addValueStrong(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
       String counterName = request.variables().get("counterName");
+      Long delta = checkForNumericParam("delta", request, responseBuilder);
+      if (delta == null) return completedFuture(responseBuilder.build());
+
+      StrongCounter strongCounter = checkForStrongCounter(counterName, responseBuilder);
+      if (strongCounter == null) return completedFuture(responseBuilder.build());
+
+      return strongCounter.addAndGet(delta).thenCompose(value -> {
```

Comment:
I think `thenApply()` fits better here. At least, you can return the `RestResponse` instead of completed future.